src/rankcorr.jl

```diff
@@ -11,209 +11,624 @@
 #######################################
 
 """
-    corspearman(x, y=x)
+    corspearman(x, y=x; skipmissing::Symbol=:none)
 
 Compute Spearman's rank correlation coefficient. If `x` and `y` are vectors, the
 output is a float, otherwise it's a matrix corresponding to the pairwise correlations
 of the columns of `x` and `y`.
+
+Uses multiple threads when either `x` or `y` is a matrix and `skipmissing` is `:pairwise`.
+
+# Keyword argument
+
+- `skipmissing::Symbol=:none`: If `:none` (the default), `missing` entries in `x` or `y`
+    give rise to `missing` entries in the return. If `:pairwise` when calculating an element
+    of the return, both `i`th entries of the input vectors are skipped if either is missing.
+    If `:listwise` the `i`th rows of both `x` and `y` are skipped if `missing` appears in
+    either; note that this might skip a high proportion of entries. Only allowed when `x` or
+    `y` is a matrix.
 """
-function corspearman(x::AbstractVector{<:Real}, y::AbstractVector{<:Real})
-    n = length(x)
-    n == length(y) || throw(DimensionMismatch("vectors must have same length"))
-    (any(isnan, x) || any(isnan, y)) && return NaN
-    return cor(tiedrank(x), tiedrank(y))
+function corspearman(x::AbstractVector, y::AbstractVector; skipmissing::Symbol=:none)
+    check_rankcor_args(x, y, skipmissing, false)
+    if x === y
+        return corspearman(x)
+    else
+        return corspearman_kernel!(x, y, skipmissing)
+    end
 end
 
-function corspearman(X::AbstractMatrix{<:Real}, y::AbstractVector{<:Real})
-    size(X, 1) == length(y) ||
-        throw(DimensionMismatch("X and y have inconsistent dimensions"))
-    n = size(X, 2)
-    C = Matrix{Float64}(I, n, 1)
-    any(isnan, y) && return fill!(C, NaN)
-    yrank = tiedrank(y)
-    for j = 1:n
-        Xj = view(X, :, j)
-        if any(isnan, Xj)
-            C[j,1] = NaN
-        else
-            Xjrank = tiedrank(Xj)
-            C[j,1] = cor(Xjrank, yrank)
+function corspearman(x::AbstractMatrix, y::AbstractVector; skipmissing::Symbol=:none)
+    return corspearman(x, reshape(y, (length(y), 1)); skipmissing)
```

Comment:
I think the 1.0 tests are failing because of this way of passing the keyword argument. I'd be okay with bumping the minimum version for StatsBase since 1.0 is ancient but maybe it's better to just make it `skipmissing=skipmissing` here for now.

Reply:
I've done some more investigation:
Other popular packages (the compat section of Project.toml) have the following minimum versions:
1.9.4 for Statistics
1.6 for Plots, DataFrames, Revise, Optim, ForwardDiff, OhMyREPL, CSV 
1.3 for Distributions
1 for StatsBase
So upping the minimum version for StatsBase seems sensible.

But as I remarked above, `eachcol` is crucial to how `corkendall` and `corspearman` are mere wrappers to `pairwise`. `eachcol` was introduced in Julia 1.1, so I imagined that upping the minimum Julia version for StatsBase to 1.1 would suffice. But no - the code appears to require Julia 1.9.4, and as of now does not work with Julia 1.8.5. For example the method `check_vectors` fails because:

```
# Julia 1.9.4
julia> keys((eachcol([1 2;3 4]))[2])
2-element LinearIndices{1, Tuple{Base.OneTo{Int64}}}:
 1
 2
```
```
# Julia 1.8.5
julia> keys((eachcol([1 2;3 4]))[2])
ERROR: MethodError: no method matching getindex(::Base.Generator{Base.OneTo{Int64}, Base.var"#242#243"{Matrix{Int64}}}, ::Int64)
Stacktrace:
 [1] top-level scope
   @ REPL[1]:1
```
So the alternatives are:

- A. We increase the minimum version to 1.9.4, which happens to be that for Statistics, but is more modern than the current LTS.
-  B. I find a workaround so that this PR requires either no increase in the minimum version number or an increase to 1.6 which is more in line with other popular packages.

Reply:
It's now the case that increasing the minimum version to 1.6 would work. See [here](https://github.com/JuliaStats/StatsBase.jl/pull/923#issuecomment-2040711012).